odl/discr/lp_discr.py

```diff
@@ -164,56 +228,82 @@ def element(self, inp=None):
                                      '{!r}.'.format(self, inp)), err)
 
     @property
-    def grid(self):
-        """Sampling grid of the discretization mappings."""
-        return self.restriction.grid
-
-    @property
-    def shape(self):
-        """Shape of the underlying grid."""
-        return self.grid.shape
-
-    @property
-    def ndim(self):
-        """Number of dimensions."""
-        return self.grid.ndim
-
-    @property
-    def cell_size(self):
-        """Cell size of an underlying regular grid."""
-        if not isinstance(self.grid, RegularGrid):
-            raise NotImplementedError('cell size not defined for non-uniform '
-                                      'grids. Use `grid.cell_sizes()` '
-                                      'instead.')
-        csize = self.grid.stride
-        idcs = np.where(csize == 0)
-        csize[idcs] = self.domain.extent[idcs]
-        return csize
-
-    @property
-    def cell_volume(self):
-        """Cell volume of an underlying regular grid."""
-        return float(np.prod(self.cell_size))
-
-    def points(self):
-        """All points in the sampling grid."""
-        return self.grid.points(order=self.order)
-
-    @property
-    def order(self):
-        """Axis ordering for array flattening."""
-        return self._order
-
-    @property
     def interp(self):
         """Interpolation type of this discretization."""
         return self._interp
 
+    # Overrides for space functions depending on partition
+    def _inner(self, x, y):
+        """Return ``self.inner(x, y)``."""
+        if any(bdry[0] or bdry[1]
+               for bdry in self.partition.nodes_on_boundary()):
+
+            # Need to halve the boundary nodes to get correct contributions.
+            # This requires copies, unfortunately.
+            # TODO: implement without copy
+            x_cpy = x.copy()
+            on_bdry = self.partition.nodes_on_boundary()
+            apply_on_boundary(
+                x_cpy.asarray(), func=lambda x: x / 2,
+                only_once=False, which_boundaries=on_bdry)
+
+            return super()._inner(x_cpy, y)
+        else:
+            # Standard case
+            return super()._inner(x, y)
+
+    def _norm(self, x):
+        """Return ``self.norm(x)``."""
+        if any(bdry[0] or bdry[1]
```

Comment:
Benchmark on this? Feels scary with an "any" here. For example in a spiral CT geometry, the grid can be 1000 x 20 x 10000, then more than 10% of the nodes are on the boundary! Quite sure this would pretty much crash then.